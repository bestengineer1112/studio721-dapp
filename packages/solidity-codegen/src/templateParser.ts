import { Lexer, rule, state, Token } from 'language-tools';
import { TokenParameter } from 'state';
import { encodeQueryParameters } from 'utils';

const lexer = new Lexer([
  state('main', [
    rule('open', '{', {
      discard: true,
      action: { type: 'push', value: 'variable' },
    }),
    ['content', '([^{]+)'],
  ]),
  state('variable', [
    rule('close', '}', { discard: true, action: { type: 'pop' } }),
    ['identifier', '([^}]+)'],
  ]),
]);

export type TokenizerResult =
  | { type: 'success'; value: Token[] }
  | { type: 'failure'; message: string };

export function parseTemplate(string: string): TokenizerResult {
  const result = lexer.tokenize(string);

  for (const token of result) {
    if (
      token.type === 'content' &&
      token.values.some((value) => value.includes('{') || value.includes('}'))
    ) {
      return { type: 'failure', message: "Brackets {} don't match!" };
    }
  }

  return { type: 'success', value: result };
}

export function parseURITemplate(string: string): TokenizerResult {
  const result = parseTemplate(string);

  if (result.type === 'success') {
    for (const token of result.value) {
      if (
        token.type === 'identifier' &&
        token.values.some((value) => !['tokenId', 'parameters'].includes(value))
      ) {
        return {
          type: 'failure',
          message: `Unrecognized variable {${token.values[0]}}.`,
        };
      }
    }
  }

  return result;
}

export function generateURI(
  template: string,
  tokenId: number,
  tokenParameters: TokenParameter[],
): string | undefined {
  const parsed = parseURITemplate(template);

  if (parsed.type === 'failure') return;

  return parsed.value
    .map((token) => {
      switch (token.type) {
        case 'content':
          return token.values.join('');
        case 'identifier': {
          switch (token.values[0]) {
            case 'tokenId':
              return tokenId.toString();
            case 'parameters':
              if (tokenParameters.length === 0) return '';

              return (
                '?' +
                encodeQueryParameters(
                  Object.fromEntries(tokenParameters.map((p) => [p.name, 0])),
                )
              );
            default:
              return '';
          }
        }
        default:
          throw new Error('Bad token type');
      }
    })
    .join('');
}

export function populateTemplate(
  template: string,
  parameters: Record<string, string>,
): string | undefined {
  const parsed = parseTemplate(template);

  if (parsed.type === 'failure') return;

  return parsed.value
    .map((token) => {
      switch (token.type) {
        case 'content':
          return token.values.join('');
        case 'identifier':
          return parameters[token.values[0]] ?? '';
        default:
          throw new Error('Bad token type');
      }
    })
    .join('');
}

export function getBaseURI(template: string): string | undefined {
  const parsed = parseURITemplate(template);

  if (parsed.type === 'failure' || parsed.value[0].type === 'identifier')
    return '';

  return parsed.value[0].values.join('');
}
