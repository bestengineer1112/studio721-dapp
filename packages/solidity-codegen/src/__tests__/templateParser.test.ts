import { generateURI, parseURITemplate } from '../templateParser';

it('parses', () => {
  const result = parseURITemplate('hello');

  expect(result).toMatchSnapshot();
});

it('parses identifier', () => {
  const result = parseURITemplate('a {tokenId} c {parameters}');

  expect(result).toMatchSnapshot();
});

it('fails to parse', () => {
  const result = parseURITemplate('a {tokenId} c parameters}');

  expect(result).toMatchSnapshot();
});

it('generates basic uri', () => {
  const result = generateURI('hello', 0, []);

  expect(result).toEqual('hello');
});

it('generates uri with token', () => {
  const result = generateURI('hello{tokenId}', 0, []);

  expect(result).toEqual('hello0');
});

it('generates uri with token & parameters', () => {
  const result = generateURI('hello{tokenId}{parameters}', 0, [
    {
      name: 'a',
      type: 'uint256',
    },
  ]);

  expect(result).toEqual('hello0?a=0');
});
