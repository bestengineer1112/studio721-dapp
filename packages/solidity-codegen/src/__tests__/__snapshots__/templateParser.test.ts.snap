// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fails to parse 1`] = `
Object {
  "message": "Brackets {} don't match!",
  "type": "failure",
}
`;

exports[`parses 1`] = `
Object {
  "type": "success",
  "value": Array [
    Object {
      "position": Object {
        "end": 5,
        "start": 0,
      },
      "type": "content",
      "values": Array [
        "hello",
      ],
    },
  ],
}
`;

exports[`parses identifier 1`] = `
Object {
  "type": "success",
  "value": Array [
    Object {
      "position": Object {
        "end": 2,
        "start": 0,
      },
      "type": "content",
      "values": Array [
        "a ",
      ],
    },
    Object {
      "position": Object {
        "end": 10,
        "start": 3,
      },
      "type": "identifier",
      "values": Array [
        "tokenId",
      ],
    },
    Object {
      "position": Object {
        "end": 14,
        "start": 11,
      },
      "type": "content",
      "values": Array [
        " c ",
      ],
    },
    Object {
      "position": Object {
        "end": 25,
        "start": 15,
      },
      "type": "identifier",
      "values": Array [
        "parameters",
      ],
    },
  ],
}
`;
